src/artifact-downloader.ts

```diff
@@ -0,0 +1,333 @@
+   */
+  async fetchFromIPFS (
+    rootCid: string,
+    artifactVariantString: ValidArtifactVariant,
+    artifactName: ArtifactName
+  ): Promise<Uint8Array> {
+    const storePath = this.#artifactDownloadsPath(
+      artifactName,
+      artifactVariantString
+    )
+
+    // Check if the artifact already exists in the store
+    if (await this.#artifactStore.exists(storePath)) {
+      dbg(
+        `Already stored ${artifactName} artifact for variant (${artifactVariantString}) - reading from cache`
+      )
+
+      const cachedData = await this.#artifactStore.get(storePath)
+      if (!cachedData) {
+        throw new Error(`Failed to read cached artifact: ${storePath}`)
+      }
+      return cachedData
+    }
+
+    await this.#initVerifiedFetch()
+
+    if (!this.#verifiedFetch) throw new Error('Verified fetch not initialized')
+
+    const cid = CID.parse(rootCid)
+    const ipfsPath = this.#getIPFSpathForArtifactName(
+      artifactName,
+      artifactVariantString
+    )
+
+    const ipfsUrl = `ipfs://${cid.toString()}/${ipfsPath}`
+
+    dbg(`Fetching from IPFS URL: ${ipfsUrl}`)
+
+    const maxRetries = 5
+    let lastError: Error | undefined
+
+    for (let attempt = 0; attempt <= maxRetries; attempt++) {
+      try {
+        // Use verified fetch with the IPFS URL
+        const response = await this.#verifiedFetch(ipfsUrl)
+
+        if (!response.ok) {
+          const statusCode = response.status
+          // Retry on 429 (rate limit), 503/504 (service unavailable/gateway timeout)
+          if ([429, 503, 504].includes(statusCode) && attempt < maxRetries) {
+            const delay = Math.pow(2, attempt) * 1_000 // exponential backoff: 1s, 2s, 4s, 8s, 16s, 32s
+            dbg(`Fetch failed with ${statusCode}, retrying in ${delay}ms (attempt ${attempt + 1}/${maxRetries})`)
+            await new Promise(resolve => setTimeout(resolve, delay))
+            continue
+          }
+          throw new Error(`Failed to fetch: ${statusCode} ${response.statusText}`)
+        }
+
+        const arrayBuffer = await response.arrayBuffer()
+        const result = new Uint8Array(arrayBuffer)
+
+        const decompressedArtifact = this.#decompressArtifact(
+          result,
+          artifactName
+        )
+
+        // Artifact integrity is ensured by the Helia node, so explicit hash validation is unnecessary.
+        await this.#artifactStore.store(
+          this.#artifactDownloadsDir(artifactVariantString),
+          storePath,
+          decompressedArtifact
+        )
+
+        dbg(`Successfully fetched and stored ${artifactName} artifact for variant (${artifactVariantString}) - ${decompressedArtifact.length} bytes`)
+
+        return decompressedArtifact
+      } catch (error) {
+        lastError = error instanceof Error ? error : new Error(String(error))
+
+        if (attempt < maxRetries) {
+          const delay = Math.pow(2, attempt) * 1000 // exponential backoff
```

Comment:
the delay couple of lines above is `const delay = Math.pow(2, attempt) * 1_000`, here's no underscore, is that intended ?